### PR TITLE
Add e2e tests to configuration files, generate jobs

### PR DESCRIPTION
**Pull request by bbguimaraes**
POC for openshift/ci-operator-prowgen#4.

/assign @smarterclayton @stevekuznetsov

### CODE REVIEW

File: ci-operator/config/openshift/descheduler/release-3.10.yaml [hunk 1]
```diff
@@ -37,3 +37,7 @@ tests:
 - as: unit
   commands: go test ./pkg/...
   from: test-bin
+- as: e2e
+  commands: cp /tmp/admin.kubeconfig /tmp/admin.conf && make test-e2e
+  openshift_ansible_src:
```
Comment (smarterclayton) [inline ci-operator/config/openshift/descheduler/release-3.10.yaml]: Looks pretty good

**Comment by stevekuznetsov** [comment]
Can we roll this out repo by repo?

**Comment by stevekuznetsov** [comment]
Starting with ones that are already 100% generated?

File: ci-operator/jobs/openshift/origin/openshift-origin-release-3.11-postsubmits.yaml [hunk 2]
```diff
@@ -20,7 +20,6 @@ postsubmits:
         - --give-pr-author-access-to-namespace=true
         - --artifact-dir=$(ARTIFACTS)
         - --target=[images]
-        - --target=artifacts
```
Comment (stevekuznetsov) [inline ci-operator/jobs/openshift/origin/openshift-origin-release-3.11-postsubmits.yaml]: These are the types of diffs we actually want to review on the generator PRs

**Comment by bbguimaraes** [comment]
Testing with:

```python
#!/usr/bin/env python3
import os
import re
import subprocess
import sys
import yaml


JOBS_DIR = 'ci-operator/jobs'


def main():
    walk(sort_lists)
    walk(lambda x: subprocess.check_call(('git', 'add', x)))
    subprocess.check_call(
        ('git', 'commit', '--quiet', '--message', 'xxx sort'))
    templates = walk(gen_e2e)
    merged = {}
    for d in templates:
        for k, v in d.items():
            l = merged.get(k, [])
            l.extend(v)
            merged[k] = l
    template_map = {
#        'cluster-launch-e2e-azure-conformance.yaml': '',
#        'cluster-launch-e2e-azure-upgrade.yaml': '',
#        'cluster-launch-e2e-azure.yaml': '',
#        'cluster-launch-e2e-gmontero-testing.yaml': '',
#        'cluster-launch-e2e-openshift-ansible.yaml': '',
#        'cluster-launch-e2e-upgrade.yaml': '',
        'cluster-launch-e2e.yaml': 'openshift_ansible',
        'cluster-launch-installer-e2e.yaml': 'openshift_installer',
        'cluster-launch-src.yaml': 'openshift_ansible_src',
#        'master-sidecar.yaml': '',
    }
    merged = {
        k: v for k, v in merged.items()
        if k in template_map}
    for template, jobs in merged.items():
        test_type = template_map.get(template)
        for filename, org, repo, branch, name, cmd, cloud in jobs:
            test = {
                'as': name,
                'commands': cmd,
                test_type: {'target_cloud': cloud}}
            with open(filename, 'r') as f:
                y = yaml.load(f)
            y['tests'].append(test)
            with open(filename, 'w') as f:
                f.write(yaml.dump(y, default_flow_style=False))


def walk(f):
    ret = []
    for root, dirs, files in os.walk(JOBS_DIR):
        for filename in files:
            if filename == 'infra-periodics.yaml':
                continue
            if not filename.endswith('.yaml'):
                continue
            ret.append(f(os.path.join(root, filename)))
    return ret


def sort_lists(filename):
    edited = False
    with open(filename) as f:
        data = yaml.load(f)
        for job_type in data:
            if job_type == 'periodics':
                continue
            for org_repo in data[job_type]:
                org, repo = org_repo.split('/')
                for job in data[job_type][org_repo]:
                    spec = job.get('spec')
                    if spec is None:
                        continue
                    container = spec['containers'][0]
                    if not runs_ci_operator(container):
                        continue
                    mounts = container.get('volumeMounts')
                    if mounts is None:
                        continue
                    spec['volumes'] = \
                        sorted(spec['volumes'], key=lambda x: x['name'])
                    if 'args' in container:
                        container['args'] = sorted(container['args'])
                    container['env'] = \
                        sorted(container['env'], key=lambda x: x['name'])
                    container['volumeMounts'] = \
                        sorted(container['volumeMounts'], key=lambda x: x['name'])
                    edited = True
    if edited:
        with open(filename, 'w') as f:
            f.write(yaml.dump(data, default_flow_style=False))


def gen_e2e(filename):
    templates = {}
    with open(filename) as f:
        data = yaml.load(f)
        for job_type in data:
            if job_type == 'periodics':
                continue
            for org_repo in data[job_type]:
                org, repo = org_repo.split('/')
                for job in data[job_type][org_repo]:
                    branch = job.get('branches', ('master'))[0]
                    spec = job.get('spec')
                    if spec is None:
                        continue
                    container = spec['containers'][0]
                    if not runs_ci_operator(container):
                        continue
                    mounts = container.get('volumeMounts')
                    if mounts is None:
                        continue
                    cmd, cloud, focus = \
                        get_named_item(container.get('env', ()), (
                            'TEST_COMMAND', 'CLUSTER_TYPE', 'TEST_FOCUS'))
                    cmd = cmd and cmd['value']
                    cloud = cloud and cloud['value']
                    focus = focus and focus['value']
                    template = get_named_item(
                        container['volumeMounts'], ('job-definition',))
                    template = template[0]['subPath']
                    name = re.match(
                        r'(?:branch|pull)-ci-{}-{}-{}-(.*)'.format(
                            org, repo, branch),
                        job['name']).group(1)
                    l = templates.get(template, [])
                    filename = 'ci-operator/config/{}/{}/{}.yaml'.format(
                        org, repo, branch)
                    l.append(
                        (filename, org, repo, branch, name, cmd, cloud))
                    templates[template] = l
    return templates


def runs_ci_operator(container):
    return next(iter(container.get('args') or ()), None) == 'ci-operator' \
        or next(iter(container.get('command') or ()), None) == 'ci-operator'


def get_named_item(l, names):
    d = {x['name']: x for x in l}
    return [d.get(x) for x in names]


main()
```

**Comment by petr-muller** [comment]
:clap: :clap: :clap:

**Pull request merged**